ReSharper.FSharp/src/FSharp.Psi.Intentions/src/Intentions/Annotation/AnnotationUtilities.fs

```diff
@@ -0,0 +1,332 @@
+        binding.ParameterPatternsEnumerable |> Seq.forall isFullyAnnotatedPattern
+
+    let isFullyAnnotatedMemberDeclaration (memberDeclaration: IMemberDeclaration) =
+        isFullyAnnotatedReturnTypeInfo memberDeclaration.ReturnTypeInfo &&
+        memberDeclaration.ParameterPatternsEnumerable |> Seq.forall isFullyAnnotatedPattern
+
+    // let f x ... =
+    // let f<'a> = ...
+    let isFunctionBinding (binding: IBinding) =
+        binding.HasParameters ||
+        binding.HeadPattern.GetNextMeaningfulSibling() :? IPostfixTypeParameterDeclarationList
+
+    // let x = y
+    // let x = fun y -> ...
+    let isValueBinding (binding: IBinding) =
+       not (isFunctionBinding binding)
+
+module SpecifyUtil =
+
+    let private addSpaceBeforeColon forceSpaceBeforeColon (pattern: ITreeNode) =
+        if forceSpaceBeforeColon || pattern.GetSettingsStoreWithEditorConfig().GetValue(fun (key: FSharpFormatSettingsKey) -> key.SpaceBeforeColon) then
+            ModificationUtil.AddChildBefore(pattern, Whitespace()) |> ignore
+
+    let private replaceWithTypedPattern typeString (pattern: IFSharpPattern) =
+        let factory = pattern.CreateElementFactory()
+        let typeUsage = factory.CreateTypeUsage(typeString)
+        ModificationUtil.ReplaceChild(pattern, factory.CreateTypedPat(pattern, typeUsage))
+
+    let addParens (pattern: IFSharpPattern) =
+        let factory = pattern.CreateElementFactory()
+        let parenPat = factory.CreateParenPat()
+        parenPat.SetPattern(pattern) |> ignore
+        ModificationUtil.ReplaceChild(pattern, parenPat) |> ignore
+
+    let private specifyReturnTypeInfo displayContext (fcsType: FSharpType) (returnTypeInfo: IReturnTypeInfo) =
+        let typeString = fcsType.Format(displayContext)
+        let factory = returnTypeInfo.CreateElementFactory()
+        // quirk warning :D
+        // for some reason returnTypeInfo.SetReturnType is failing on a case like let x: [<A>] _ = 1
+        // and sets returnType before AttributeList resulting in something like let x: int[<A>] = 1
+        // so I guess we have to save attribute list and recreate new returnTypeInfo with it
+        let newTypeInfo = factory.CreateReturnTypeInfo(factory.CreateTypeUsage(typeString))
+
+        let mutable idx = 0
+        if not (returnTypeInfo.AttributeListsEnumerable.IsEmpty()) then
+            for attributeList in returnTypeInfo.AttributeListsEnumerable do
+                FSharpAttributesUtil.addAttributeList false newTypeInfo.ReturnType
+                let mutable prevAttribute = null
+                for attribute in attributeList.AttributesEnumerable do
+                    if isNull prevAttribute then
+                        prevAttribute <- FSharpAttributesUtil.addAttribute newTypeInfo.AttributeLists[idx] attribute
+                    else
+                        FSharpAttributesUtil.addAttributeAfter prevAttribute attribute
+                idx <- idx + 1
```

Comment:
Why does it need replacing the whole return type info instead of replacing the type?